src/knot_protocol.c

```diff
@@ -95,80 +95,75 @@
-	{ KNOT_TYPE_ID_PRESSURE,		VTYPE_INT, UNIT_PRESSURE_MASK	},
-	{ KNOT_TYPE_ID_DISTANCE,		VTYPE_INT, UNIT_DISTANCE_MASK	},
-	{ KNOT_TYPE_ID_ANGLE,			VTYPE_FLOAT, UNIT_ANGLE_MASK	},
-	{ KNOT_TYPE_ID_VOLUME,			VTYPE_FLOAT, UNIT_VOL_MASK	},
-	{ KNOT_TYPE_ID_AREA,			VTYPE_FLOAT, UNIT_AREA_MASK	},
-	{ KNOT_TYPE_ID_RAIN,			VTYPE_FLOAT, UNIT_RAIN_MASK	},
-	{ KNOT_TYPE_ID_DENSITY,			VTYPE_FLOAT, UNIT_DENSITY_MASK	},
-	{ KNOT_TYPE_ID_LATITUDE,		VTYPE_FLOAT, UNIT_LAT_MASK	},
-	{ KNOT_TYPE_ID_LONGITUDE,		VTYPE_FLOAT, UNIT_LONG_MASK	},
-	{ KNOT_TYPE_ID_SPEED,			VTYPE_INT,	UNIT_SPEED_MASK },
-	{ KNOT_TYPE_ID_VOLUMEFLOW,		VTYPE_FLOAT, UNIT_VOLFLOW_MASK	},
-	{ KNOT_TYPE_ID_ENERGY,			VTYPE_INT, UNIT_ENERGY_MASK	},
-
-	{ KNOT_TYPE_ID_PRESENCE,		VTYPE_BOOL, 0xFFFF		},
-	{ KNOT_TYPE_ID_SWITCH,			VTYPE_BOOL, 0xFFFF		},
-	{ KNOT_TYPE_ID_COMMAND,			VTYPE_RAW, 0xFFFF		},
-	{ KNOT_TYPE_ID_RESERVED,		VTYPE_RAW, 0xFFFF		},
+} basic_types[]  = {
+	{KNOT_VALUE_TYPE_RAW, 0xFFFF			},
+	{KNOT_VALUE_TYPE_INT, UNIT_VOLTAGE_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_CURRENT_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_RESISTENCE_MASK	},
+	{KNOT_VALUE_TYPE_INT, UNIT_POWER_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_TEMP_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_HUM_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_LUM_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_TIME_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_MASS_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_PRESSURE_MASK	},
+	{KNOT_VALUE_TYPE_INT, UNIT_DISTANCE_MASK	},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_ANGLE_MASK		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_VOL_MASK		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_AREA_MASK		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_RAIN_MASK		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_DENSITY_MASK	},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_LAT_MASK		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_LONG_MASK		},
+	{KNOT_VALUE_TYPE_INT, UNIT_SPEED_MASK 		},
+	{KNOT_VALUE_TYPE_FLOAT, UNIT_VOLFLOW_MASK	},
+	{KNOT_VALUE_TYPE_INT, UNIT_ENERGY_MASK		}
+};
+
+uint8_t logic_types[]  = {
+	KNOT_VALUE_TYPE_BOOL,			// KNOT_TYPE_ID_PRESENCE
+	KNOT_VALUE_TYPE_BOOL,			// KNOT_TYPE_ID_SWITCH
+	KNOT_VALUE_TYPE_RAW 			// KNOT_TYPE_ID_COMMAND
 };
 
 int knot_value_type_is_valid(uint8_t type)
 {
 	if (type < KNOT_VALUE_TYPE_MIN || type > KNOT_VALUE_TYPE_MAX)
-		return ERANGE;
+		return KNOT_INVALID_DATA;
 
-	return 0;
+	return KNOT_SUCCESS;
 }
 
-int knot_schema_is_valid(uint8_t type_id, uint8_t value_type, uint8_t unit)
+int knot_schema_is_valid(uint16_t type_id, uint8_t value_type, uint8_t unit)
 {
-	int i;
-
 	/* int/float/bool/raw ? */
-	if (knot_value_type_is_valid(value_type))
-		return EINVAL;
-
-	/* Not expecting more than 32 diff unit for the same type */
-	if (unit > 31)
-		return EINVAL;
-
-	for (i = 0; schema[i].type_id != KNOT_TYPE_ID_RESERVED; i++) {
-
-		if (schema[i].type_id != type_id)
-			continue;
-
-		if (schema[i].value_type != value_type)
-			return EINVAL;
-
-		if (schema[i].unit_mask & UNIT_MASK(unit))
-			return 0;
-		else
-			return EINVAL;
+	if (knot_value_type_is_valid(value_type) == KNOT_SUCCESS) {
```

Comment:
I recommend to invert the logic and return the same error returned by knot_value_type_is_valid() instead of map all errors to "invalid schema"